docs/source/config.rst
Configuration Options
#####################

Enable openNDS
**************

Set to 0 to disable opennds

``option enabled 1``

Use deprecated generic configuration file
******************************************

Use of this setting is no longer supported and if present is ignored.

``option config '/etc/opennds/opennds.conf'``

Enable debug output (0-3)
*************************

Default: 1

Level0
------
Silent (only initial startup, LOG_ERR and LOG_EMERG messages will be seen, otherwise there will be no logging.)

Level 1
-------
LOG_ERR, LOG_EMERG, LOG_WARNING and LOG_NOTICE (this is the default level).

Level 2
-------
debuglevel 1  + LOG_INFO

Level 3
-------
debuglevel 2 + LOG_DEBUG

``option debuglevel '1'``

Firewall Restart hook
*********************

Specific to OpenWrt only, Firewall 4 (FW4) informs openNDS when it is restarting.

If enabled (Set to 1), openNDS reinserts any nftables it may need in the FW4 ruleset.

Default: 1

``option fwhook_enabled '1'``

Ndsctl Socket Access
********************

Set the socket name to use for ndsctl socket access, relative to the tmpfs mountpoint.

Any directory/folder specified must exist.
Default: ndsctl.sock (Do not add a leading "/")

Full default socket path would be /tmp/ndsctl.sock in OpenWrt
In the following example, the socket path would be /tmp/sockets/ndsctl.sock

Example:

``option ndsctlsocket 'sockets/ndsctl.sock'``

Local Log Mountpoint
********************

Default: router's volatile tmpfs storage eg on OpenWrt '/tmp'

Local logging can be directed to any storage accessible to the router eg USB drive, SSD etc

**WARNING** - you should not use the router's built in flash storage as this would cause
excessive wear and in a live system will result quite quickly in flash failure making the router useless.

Example:

``option log_mountpoint '/logs'``

Maximum number of Local Log Entries
***********************************

Set the maximum number of local log entries to be kept.
Default 100

Minimum value 0 (no limit)

Maximum value - limited only by free storage space on the logging mountpoint

If set to '0' there is no limit

This is the maximum number of local log entries allowed before log rotation begins

Both ThemeSpec and BinAuth log locally if they are enabled

**WARNING** - local logging is by default written to the tmpfs volatile storage

If this option were to be set too high the router could run out of tmpfs storage and/or free RAM

Non-volatile storage, such as a USB storage device may be defined using the log_mountpoint option

Example:

``option max_log_entries '1000'``

Login Option
************

Default: 1

Integer value sent to PreAuth script as login mode

openNDS comes preconfigured for three basic modes of operation

Mode 0
------
If FAS is not enabled, then this functions as mode 1

If FAS is configured, FAS is used.

Mode 1
------
Default Dynamic Click to Continue

The pre-installed dynamic login page is enabled by setting option login_option_enabled = '1'

It generates a Click to Continue page followed by an info/advertising page.

User clicks on “Continue” are recorded in the log file /[tmpfs_dir]/ndslog/ndslog.log

Mode 2
------
Username/Emailaddress Dynamic Login

The pre-installed dynamic login page is enabled by setting option login_option_enabled = '2'

It generates a login page asking for username and email address followed by an info/advertising page.

User logins are recorded in the log file /[tmpfs_dir]/ndslog/ndslog.log

Mode 3
------
Use Theme defined in ThemeSpec path (option themespec_path)

`option login_option_enabled '1'`

Allow Preemptive Authentication
*******************************

Default: 1 - Enabled

Disable by setting to 0

This allows the ndsctl utility to preemptively authorise **connected** clients that have not entered the preauthenticated state.

This is useful for example with IoT devices that do not have CPD (captive portal detection)

or for a FAS to manage inter-captive-portal roaming by making use of a centralised database of client validations.

Example:

``option allow_preemptive_authentication '0'``

ThemeSpec Path
**************

Default: None

Required when when login_option_enabled is set to '3'

Note: /usr/lib/opennds/theme_click-to-continue-basic.sh is used for login_option_enabled '1'

and:  /usr/lib/opennds/theme_user-email-login-basic.sh is used for login_option_enabled '2'

Sets the ThemeSpec file path to be used when login_option_enabled '3'

The ThemeSpec script makes use of lists of custom parameters, custom variables, custom image urls and custom files and is used to generate the dynamic splash page sequence.

The ThemeSpec file will normally reside in /usr/lib/opennds/ but can be anywhere accessible to openNDS.

The file must be flagged as executable and have the correct shebang for the default shell.

``option themespec_path '/usr/lib/opennds/<filename>'``

DHCP Leases File
****************

Default: Try /tmp/dhcp.leases, /var/lib/misc/dnsmasq.lease or /var/db/dnsmasq.leases

The file containing the list of active DHCP leases.

Example:

``option dhcp_leases_file '/tmp/dhcp.leases.special'``

Set refresh interval for downloads
**********************************

Set refresh interval for downloaded remote files (in minutes)

Default 0

A setting of 0 (zero) means refresh is disabled.

This is useful for providing automated refreshing of informational or advertising content. Should the remote resources become unavailable, current versions will continue to be used.

Example, set to twelve hours (720 minutes):

``option remotes_refresh_interval '720'``

Use outdated libmicrohttpd (MHD)
********************************

Default 0 (Disabled)

**Warning**: enabling this *may* cause instability or in the worst case total failure - it would be better to upgrade MHD.

**Use at your own risk.**

Older versions of MHD use an older version of the MHD API and may not run correctly or fail.

Older versions of MHD convert & and + characters to spaces when present in form data. This can make a PreAuth or BinAuth impossible to use for a client if form data contains either of these characters eg. in username or password.

*There may well be other issues with older versions.*

MHD versions earlier than 0.9.71 are detected.

If this option is set to 0 (default), NDS will terminate if MHD is earlier than 0.9.71

If this option is set to 1, NDS will attempt to start and log an error.

``option use_outdated_mhd '1'``

Maximum Page Size to be served by MHD
*************************************

Default 10240 bytes

Minimum value 1024 bytes

Maximum - limited only by free RAM in the router

This sets the maximum number of bytes that will be served per page by the MHD web server.

Setting this option is useful:

	1. To reduce memory requirements on a resource constrained router
	2. To allow large pages to be served where memory usage is not a concern

Example:

``option max_page_size '4096'``

Set the MHD WebRoot
*******************

Default: /etc/opennds/htdocs

The local path where the system css file, and other static page content resides.

ie. Serve the file splash.css from this directory

Example:

``option webroot '/etc/opennds/htdocs'``

Set the GatewayInterface
************************

Default: br-lan

Use this option to set the device opennds will bind to.

The value may be an interface section in /etc/config/network or a device name such as br-lan.

The selected interface must be allocated an IPv4 address.

In OpenWrt this is normally br-lan, in generic Linux it might be wlan0

``option gatewayinterface 'br-lan'``

Set the GatewayPort
*******************

Default: 2050

openNDS's own http server (MHD) uses the gateway address as its IP address.

This option sets the port it listens to.

Example:

``option gatewayport '2080'``

Set the GatewayName
*******************

Default: openNDS

gatewayname is used as an identifier for the instance of openNDS

It is displayed on the default splash page sequence for ThemeSpec and the example php scripts.

It is particularly useful in the case of a single remote FAS server that serves multiple openNDS sites, allowing the FAS to customise its response for each site.

Note: The single quote (or apostrophe) character ('), cannot be used in the gatewayname.

If it is required, use the htmlentity &#39; instead.

For example:

``option gatewayname 'Bill's WiFi'`` is invalid.

Instead use:

``option gatewayname 'Bill&#39;s WiFi'``

Example:

``option gatewayname 'OpenWrt openNDS'``

Serial Number Suffix Enable
***************************

Appends a serial number suffix to the gatewayname string.

openNDS constructs a serial number based on the router mac address and adds it to the gatewayname

Default: 1 (enabled)

To disable, set to 0

Example:

``option enable_serial_number_suffix '0'``

Set GatewayFQDN
***************

Default: status.client

This is the simulated FQDN used by a client to access the Client Status Page

If not set, the Status page can be accessed at: http://gatewayaddress:gatewayport/

Warning - if set, services on port 80 of the gateway will no longer be accessible (eg Luci AdminUI)

By default, the Error511/Status page will be found at http://status.client/ by a redirection of port 80 to http://gatewayaddress:gatewayport/

Disable GatewayFQDN by setting the option to 'disable'

ie:

``option gatewayfqdn 'disable'``

Alternate Useful Example:

``option gatewayfqdn 'login.page'``

Set StatusPath
**************

Default: /usr/lib/opennds/client_params.sh

This is the script used to generate the GatewayFQDN client status page.

Example:

``option statuspath '/mycustomscripts/custom_client_params.sh'``

Set MaxClients
**************

Default 250

The maximum number of clients allowed to connect.

**This should be less than or equal to the number of allowed DHCP leases.** set for the router's dhcp server.

Example:

``option maxclients '500'``

Set Fast Client Scan
********************

Default 0 (disabled)

Enables fast client scanning to find the interface used by the client.

Fast scanning skips search for actual wireless interface used.

Any scan delay effects the display rewrite speed of the splash page sequence.

Useful when a large mesh or other extended network is in use.

Otherwise the Client Scan could, for large networks, take many seconds to complete.

For example, to enable:
``option fast_client_scan '1'``

Client timeouts in minutes
**************************

Preauthidletimeout
------------------

Default 30

This is the time in minutes after which a client is disconnected if not authenticated.

ie the client has not attempted to authenticate for this period.

Example:

``option preauthidletimeout '60'``

Authidletimeout
---------------

Default 120

This is the time in minutes after which an idle client is disconnected
ie the client has not used the network access for this period

Example:

``option authidletimeout '60'``

Session Timeout
---------------

Default 1440 minutes (24 hours).

This is the interval after which clients are forced out (a value of 0 means never).

Clients will be deauthenticated at the end of this period.

Example: Set to 20 hours (1200 minutes).

``option sessiontimeout '1200'``

Set the Checkinterval
*********************

The interval in seconds at which openNDS checks client timeouts, quota usage and runs watchdog checks.

Default: 15 seconds (one quarter of a minute).

Example: Set to 30 seconds.

``option checkinterval '30'``

Set Rate Quotas
***************

Defaults 0

Integer values only.

.. note::
 Upload means *to* the Internet, download means *from* the Internet.

If the client average data rate exceeds the value set here, the client will be rate limited.

Values are in kb/s.

If set to 0, there is no limit.

Quotas and rates can also be set by FAS via Authmon Daemon, ThemeSpec scripts, BinAuth, and ndsctl auth. Values set by these methods, will override values set in the config file.

Rates:

``option uploadrate '200'``

``option downloadrate '800'``

Set Bucket Ratio
****************

Default 10

Upload and Download bucket ratios can be defined.

Allows fine control of upload rate limit threshold overrun per client.

Used in conjunction with MaxDownloadBucketSize and MaxUploadBucketSize.

Facilitates calculation of a dynamic "bucket size" or "queue length" (in packets) to be used for buffering upload and download traffic to achieve rate restrictions defined in this config file or by FAS for individual clients.

If a bucket becomes full, packets will overflow and be dropped to maintain the rate limit.

To minimise the number of dropped packets the bucket ratio can be increased whilst still maintaining the configured rate restriction.

***CAUTION*** Large values may consume large amounts of memory per client.

If the client's average rate does not exceed its configured value within the ratecheck window interval (See RateCheckWindow option), no memory is consumed.

If the rate is set to 0, the Bucket Ratio setting has no meaning and no memory is consumed.

Examples:

	`option upload_bucket_ratio '1'`

	`option download_bucket_ratio '5'`


MaxDownloadBucketSize
*********************

Default: 250

Allows control over download rate limiting packet loss at the expense of increased latency.

***CAUTION*** Large values may consume large amounts of memory per client.

Allowed Range 5 to 10000

Example:

`option max_download_bucket_size '100'`

MaxUploadBucketSize
*******************

Default 250

Allows control over upload rate limiting packet loss at the expense of increased latency.


ChangeLog
opennds (10.3.0)
  * Fix - missing exit in library call [bluewavenet]
  * Add - support for mesh11sd meshnodes and reimplement auth_restore not to use ndsctl -  [bluewavenet]
  * Fix - formatting of ThemeSpec scripts -  [bluewavenet]
  * Fix - move parsing of custom vaiables and fields to libopennds library script -  [bluewavenet]
  * Fix - move custom_variables_list parsing to liboppennds library -  [bluewavenet]
  * Fix - parsing of custom variables list -  [bluewavenet]
  * Fix - non placeholder themespec logo size -  [bluewavenet]
  * Fix - formatting of themespec custom placeholders and status page -  [bluewavenet]
  * Fix - missing parseing of custom files list in config -  [bluewavenet]
  * Fix - issue #596, Problem with fasremotefqdn that has multiple IPs ( CDN ) -  [bluewavenet] [viniciussotti ]
  * Add - download protocol debug message -  [bluewavenet]
  * Add - function resolve_fqdn() -  [bluewavenet]
  * Fix - failure to parse custom variables and custom images from config lists -  [bluewavenet]
  * Fix - issue #582, ensure list dhcp-option-force is added to the correct dnsmasq zone -  [bluewavenet] [flintt]
  * Fix - issue #591, error adding nft rule when fasremotefqdn not set -  [bluewavenet] [billz]
  * Fix - change error message to debug message when legacy ipset not present -  [bluewavenet]
  * Fix - issue #588, Remove default value for faskey -  [bluewavenet] [billz]
  * Fix - some memory leaks -  [bluewavenet] [LuMingYinDetect]
  * Fix - typo preventing reload of dnsmasq -  [bluewavenet]
  * Fix - issue #571, CVE-2024-25763, use after free - we now delete the client instead of changing state -  [bluewavenet] [LuMingYinDetect]
  * Fix - Evaluate system call return values -  [bluewavenet] [sunweaver]
  * Fix - enable iteration counter for all nft error codes -  [bluewavenet]
  * Fix - issue #556, use dnsmasq reload instead of restart -  [bluewavenet] [rafaelfelix]
  * Fix - remove unneeded memory allocation -  [bluewavenet]
  * Fix - ensure all client memory is freed when removing client -  [bluewavenet]
  * Fix - remove unneeded memory allocations and free some strdup allocations -  [bluewavenet]
  * Fix - race condition when dnsmasq takes too long to restart -  [bluewavenet]
  * Add - issue #547, Support specifying DHCP leases file location [mcassaniti]

 -- Rob White <dot@blue-wave.net>  Wed, 07 Jun 2024 07:31:46 +0000

opennds (10.2.0)
  * Add - Page 202 HTTP_ACCEPTED for future use [bluewavenet]
  * Fix - remove redundant workaround for old MHD versions [bluewavenet]
  * Fix - some nft and other error messages [bluewavenet]
  * Fix - remove unnecessary debug messages [bluewavenet]
  * Add - improved ndsctl status detection for authmon [bluewavenet]
  * Add - increase RestartSec parameter in opennds.service for generic Linux [bluewavenet]
  * Fix - prevent unnecessary shutdown [bluewavenet]
  * Fix - Generic Linux, error updating dnsmasq.conf [bluewavenet]
  * Add - allow dynamic update of flowtable rules [bluewavenet]
  * Fix - use Themespec in place of deprecated preauth in ndectl status [bluewavenet]
  * Fix - Generic linux -  keep old config [bluewavenet]
  * Fix - remove some unused variables [bluewavenet]
  * Add - support for nftables blocklists [bluewavenet]
  * Add - ensure authenticated user rules are added in list order [bluewavenet]
  * Add - Set default authenticated policy to accept [bluewavenet]
  * Add - urandom hash to key generation [bluewavenet]
  * Fix - Fix - duplicate users_to_router rules [bluewavenet]
  * Add - Automatic dns resolution of fas_remotefqdn in nftables rules [bluewavenet]
  * Add - flowtables rules [bluewavenet]
  * Add - dynamic flowtable support allowing multiple upstream connections [bluewavenet]
  * Add - skip preemptivemac client if not dhcp database or is already authenticated [bluewavenet]
  * Add - Skip auth_restore if client is in preemptivemac list [bluewavenet]
  * Add - use daemon_auth in auth_restore [bluewavenet]
  * Add - Dynamic refresh of configured preemptive macs [bluewavenet]
  * Fix - suppress demon_auth debug output [bluewavenet]
  * Add - urlencode ALL list blocks and introduce preemptivemac lists [bluewavenet]
  * Add - fas-hid-https to makefiles [bluewavenet]
  * Add - warning that pre-shared key will be generated and added to config if not present [bluewavenet]
  * Add - b64decode payload in fas [bluewavenet]
  * Add - b64encode payload before sending to fas [bluewavenet]
  * Add - level 4 fas-hid-https [bluewavenet]
  * Add - support for fas_secure_enabled = 4 [bluewavenet]
  * Add - updates to comments in fas-hid script [bluewavenet]
  * Fix - fas-hid icon position [bluewavenet]
  * Add - wget_request support to authmon [bluewavenet]
  * Add - ruleset full parsing of verdict, protocol ports to/from address [bluewavenet]
  * Fix - send_to_fas_deauthed [bluewavenet]
  * Fix - ensure action is parsed correctly in all cases [bluewavenet]
  * Add - Quota based Fair Usage Policy, sets throttled rate when quota exceeded [bluewavenet]
  * Add - QL code scanning support in Community theme_voucher ThemeSpec script [bluewavenet]
  * Add - support for cpi_query in example FAS scripts [bluewavenet]
  * Fix - memory leak when deleting client from client list [bluewavenet]
  * Add - html entity handling for semicolon [bluewavenet]
  * Add - Store RFC8910 request string in client data [bluewavenet]

 -- Rob White <dot@blue-wave.net>  Wed, 22 Nov 2023 11:08:15 +0000

opennds (10.1.3)
  * Security Advisory. This version contains fixes for multiple potential security vulnerabilities
  * Credit - Stanislav Dashevskyi - standash.github.io [standash]
  * It also contains some minor bug fixes
  * Fix - Buffer overflow causing segfault - CVE-2023-41101 [bluewavenet]
  * Fix - Memory leaks due to passing allocated buffer into safe_asprintf() - CVE-2023-41102 [bluewavenet]
  * Fix - Remove deprecated preauth option [bluewavenet]
  * Fix - missing free in show_preauth_page if MHD does not respond [bluewavenet]

src/conf.h
/********************************************************************\
 * This program is free software; you can redistribute it and/or    *
 * modify it under the terms of the GNU General Public License as   *
 * published by the Free Software Foundation; either version 2 of   *
 * the License, or (at your option) any later version.              *
 *                                                                  *
 * This program is distributed in the hope that it will be useful,  *
 * but WITHOUT ANY WARRANTY; without even the implied warranty of   *
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the    *
 * GNU General Public License for more details.                     *
 *                                                                  *
 * You should have received a copy of the GNU General Public License*
 * along with this program; if not, contact:                        *
 *                                                                  *
 * Free Software Foundation           Voice:  +1-617-542-5942       *
 * 59 Temple Place - Suite 330        Fax:    +1-617-542-2652       *
 * Boston, MA  02111-1307,  USA       gnu@gnu.org                   *
 *                                                                  *
\********************************************************************/

/** @file conf.h
    @brief Config file parsing
    @author Copyright (C) 2004 Philippe April <papril777@yahoo.com>
    @author Copyright (C) 2007 Paul Kube <nodogsplash@kokoro.ucsd.edu>
    @author Copyright (C) 2015-2025 Modifications and additions by BlueWave Projects and Services <opennds@blue-wave.net>
*/

#define COPYRIGHT "openNDS, Copyright (C) 2015-2025 Modifications and additions by BlueWave Projects and Services"

#ifndef _CONF_H_
#define _CONF_H_

#define VERSION "10.3.2beta"

/*
 * Defines how many times should we try detecting the interface with the default route (in seconds).
 * If set to 0, it will keep retrying forever.
 */
#define NUM_EXT_INTERFACE_DETECT_RETRY 0

// How long we should wait per try to detect the interface with the default route if it isn't up yet (interval in seconds)
#define EXT_INTERFACE_DETECT_RETRY_INTERVAL 1

// Defaults configuration values
#ifndef SYSCONFDIR
#define DEFAULT_CONFIGFILE "/etc/opennds/opennds.conf"
#else
#define DEFAULT_CONFIGFILE SYSCONFDIR"/opennds/opennds.conf"
#endif
#define DEFAULT_DAEMON "0"
#define DEFAULT_ENABLED "1"
#define DEFAULT_DEBUGLEVEL "1"
#define DEFAULT_MAXCLIENTS "250"
#define DEFAULT_ONLINE_STATUS "0"
#define DEFAULT_GATEWAYINTERFACE "br-lan"
#define DEFAULT_GATEWAY_IPRANGE "0.0.0.0/0"
#define DEFAULT_GATEWAYNAME "openNDS"
#define DEFAULT_ENABLE_SERIAL_NUMBER_SUFFIX "1"
#define DEFAULT_GATEWAYPORT "2050"
#define DEFAULT_GATEWAYFQDN "status.client"
#define DEFAULT_DHCP_DEFAULT_URL_ENABLE "1"
#define DEFAULT_STATUSPATH "/usr/lib/opennds/client_params.sh"
#define DEFAULT_LOG_MOUNTPOINT "/tmp"
#define DEFAULT_MAX_PAGE_SIZE "10240"
#define DEFAULT_FASPORT "0"
#define DEFAULT_LOGIN_OPTION_ENABLED "0"
#define DEFAULT_MAX_LOG_ENTRIES "100"
#define DEFAULT_USE_OUTDATED_MHD "0"
#define DEFAULT_ALLOW_PREEMPTIVE_AUTHENTICATION "1"
#define DEFAULT_FAS_SECURE_ENABLED "1"
#define DEFAULT_FASPATH "/"
#define DEFAULT_FASKEY ""
#define DEFAULT_BINAUTH "/usr/lib/opennds/binauth_log.sh"
#define DEFAULT_CHECKINTERVAL "15"
#define DEFAULT_SESSION_TIMEOUT "1440"
#define DEFAULT_PREAUTH_IDLE_TIMEOUT "30"
#define DEFAULT_AUTH_IDLE_TIMEOUT "120"
#define DEFAULT_REMOTES_REFRESH_INTERVAL "0"
#define DEFAULT_WEBROOT "/etc/opennds/htdocs"
#define DEFAULT_TMPFSMOUNTPOINT "/tmp"
#define DEFAULT_AUTHDIR "opennds_auth"
#define DEFAULT_DENYDIR "opennds_deny"
#define DEFAULT_PREAUTHDIR "opennds_preauth"
#define DEFAULT_SET_MSS "1" //allow setting the TCP Maximum Segment Size
#define DEFAULT_MSS_VALUE "0" // value to set the MSS. 0 means use max possible ie clamp-mss-to-pmtu
#define DEFAULT_RATE_CHECK_WINDOW "2" // The data rate check moving average window size multiply this by CHECKINTERVAL to give window size (or burst interval) in seconds
#define DEFAULT_UPLOAD_RATE "0" // 0 means no limit
#define DEFAULT_DOWNLOAD_RATE "0" // 0 means no limit
#define DEFAULT_UPLOAD_BUCKET_RATIO "1" // Allows control of upload rate limit threshold overrun per client
#define DEFAULT_DOWNLOAD_BUCKET_RATIO "1" // Allows control of download rate limit threshold overrun per client
#define DEFAULT_MAX_UPLOAD_BUCKET_SIZE "250" // Allows control over upload rate limiting packet loss at the expense of increased latency
#define DEFAULT_MAX_DOWNLOAD_BUCKET_SIZE "250" // Allows control over download rate limiting packet loss at the expense of increased latency
#define DEFAULT_UPLOAD_QUOTA "0" // 0 means no limit
#define DEFAULT_DOWNLOAD_QUOTA "0" // 0 means no limit
#define DEFAULT_FUP_UPLOAD_THROTTLE_RATE "0" // 0 means BLOCK the client until deauthed
#define DEFAULT_FUP_DOWNLOAD_THROTTLE_RATE "0" // 0 means BLOCK the client until deauthed
#define DEFAULT_UPLOAD_UNRESTRICTED_BURSTING "0" // 0 means disabled, 1 means enabled
#define DEFAULT_DOWNLOAD_UNRESTRICTED_BURSTING "0" // 0 means disabled, 1 means enabled
#define DEFAULT_NDSCTL_SOCK "ndsctl.sock"
#define DEFAULT_FW_MARK_AUTHENTICATED "0x30000"
#define DEFAULT_FW_MARK_AUTH_BLOCKED "0x30001"
#define DEFAULT_AUTHENTICATION_MARK "0x00030000"
#define DEFAULT_FW_MARK_TRUSTED "0x20000"
#define DEFAULT_THEMESPEC_PATH ""
#define DEFAULT_FAS_REMOTEFQDN "disabled"
#define DEFAULT_FAS_REMOTEIP "disabled"
#define DEFAULT_FAS_SSL "wget"

/* N.B.: default policies here must be ACCEPT, REJECT, or RETURN
 * In the .conf file, they must be allow, block, or passthrough
 * Mapping between these enforced by parse_empty_ruleset_policy()
 */
#define DEFAULT_EMPTY_TRUSTED_USERS_POLICY "accept"
#define DEFAULT_EMPTY_TRUSTED_USERS_TO_ROUTER_POLICY "accept"
#define DEFAULT_EMPTY_USERS_TO_ROUTER_POLICY "reject"
#define DEFAULT_EMPTY_AUTHENTICATED_USERS_POLICY "return"
#define DEFAULT_EMPTY_PREAUTHENTICATED_USERS_POLICY "reject"
#define DEFAULT_IP6 0

// Default lists
#define DEFAULT_TRUSTEDMACLIST ""
#define DEFAULT_FAS_CUSTOM_PARAMETERS_LIST ""
#define DEFAULT_FAS_CUSTOM_VARIABLES_LIST ""
#define DEFAULT_FAS_CUSTOM_IMAGES_LIST ""
#define DEFAULT_FAS_CUSTOM_FILES_LIST ""
#define DEFAULT_USERS_TO_ROUTER "allow%20udp%20port%2053 allow%20udp%20port%2067 allow%20tcp%20port%2022 allow%20tcp%20port%20443"
#define DEFAULT_AUTHENTICATED_USERS "allow%20all"
#define DEFAULT_PREAUTHENTICATED_USERS ""

// Firewall targets
typedef enum {
	TARGET_DROP,
	TARGET_REJECT,
	TARGET_ACCEPT,
	TARGET_RETURN,
	TARGET_LOG,
	TARGET_ULOG
} t_firewall_target;

// Firewall rules
typedef struct _firewall_rule_t {
	t_firewall_target target;	//@brief t_firewall_target
	char *protocol;		//@brief tcp, udp, etc ...
	char *port;			//@brief Port to block/allow
	char *mask;			//@brief Mask for the rule *destination*
	char *ipset;			//@brief IPset rule
	struct _firewall_rule_t *next;
} t_firewall_rule;

// MAC Addresses
typedef struct _MAC_t {
	char *mac;
	struct _MAC_t *next;
} t_MAC;

// Walled Garden Ports
typedef struct _WGP_t {
	unsigned int wgport;
	struct _WGP_t *next;
} t_WGP;

// Walled Garden FQDNs
typedef struct _WGFQDN_t {
	char *wgfqdn;
	struct _WGFQDN_t *next;
} t_WGFQDN;

// Custom FAS Parameters
typedef struct _FASPARAM_t {
	char *fasparam;
	struct _FASPARAM_t *next;
} t_FASPARAM;

// Custom FAS Variables
typedef struct _FASVAR_t {
	char *fasvar;
	struct _FASVAR_t *next;
} t_FASVAR;

// Custom FAS Images
typedef struct _FASIMG_t {
	char *fasimg;
	struct _FASIMG_t *next;
} t_FASIMG;

// Custom FAS Files
typedef struct _FASFILE_t {
	char *fasfile;
	struct _FASFILE_t *next;
} t_FASFILE;

// Configuration structure
typedef struct {
	char configfile[255];					//@brief name of the config file
	char *ndsctl_sock;					//@brief ndsctl path to socket
	char *internal_sock;					//@brief internal path to socket
	int enabled;						//@brief if openNDS is enabled
	int daemon;						//@brief if daemon != 0, use daemon mode
	int debuglevel;						//@brief Debug information verbosity
	int maxclients;						//@brief Maximum number of clients allowed
	int online_status;					//@brief Online status of the router, 1=online, 0=offline
	char *gw_name;						//@brief Name of the gateway; e.g. its SSID or a unique identifier for use in a remote FAS
	int enable_serial_number_suffix;			//@brief Enable/disable serial number suffix to gateway name
	char *http_encoded_gw_name;				//@brief http encoded name of the gateway, used as a templated variable in splash.htm
	char *url_encoded_gw_name;				//@brief url encoded name of the gateway used as variable in Preauth
	char *gw_interface;					//@brief Interface we will manage
	char *ext_gateway;					//@brief The interfaces and IP addresses of upstream gateways
	char *gw_iprange;					//@brief IP range on gw_interface we will manage
	char *gw_ip;						//@brief Internal IP (v4 or v6) for our web server
	char *gw_address;					//@brief Internal IP with port for our web server
	char *gw_mac;						//@brief MAC address of the interface we manage
	char *gw_fqdn;						//@brief FQDN of the client status page
	char *status_path;					//@brief Path to the client status page script
	int dhcp_default_url_enable;				//@brief Enable DHCP default-url (code 114 - RFC8910)
	unsigned int gw_port;					//@brief Port the webserver will run on
	unsigned int fas_port;					//@brief Port the fas server will run on
	int login_option_enabled;				//@brief Use default PreAuth Login script
	unsigned long long int max_log_entries;			//@brief set the maximum number of log entries
	int use_outdated_mhd;					//@brief Use outdated libmicrohttpd
	unsigned long long int max_page_size;			//@brief Max page size to be served by libmicrohttpd
	int allow_preemptive_authentication;			//@brief Allow Preemptive Authentication using the ndsctl utility
	int fas_secure_enabled;					//@brief Enable Secure FAS
	char *fas_path;						//@brief Path to forward authentication page of FAS
	char *fas_key;						//@brief AES key for FAS
	char *fas_remoteip;					//@brief IP addess of a remote FAS
	char *fas_remotefqdn;					//@brief FQDN of a remote FAS
	char *fas_url;						//@brief URL of a remote FAS
	char *fas_ssl;						//@brief SSL provider for FAS
	char *fas_hid;						//@brief Hash provider for FAS
	char *themespec_path;					//@brief Path to the ThemeSpec file to use for login_option_enabled = 3
	char *tmpfsmountpoint;					//@brief Mountpoint of the tmpfs drive eg /tmp etc.
	char *log_mountpoint;					//@brief Mountpoint of the log drive eg a USB drive mounted at /logs
	char *webroot;						//@brief Directory containing splash pages, etc.
	char *authdir;						//@brief Notional relative dir for authentication URL
	char *denydir;						//@brief Notional relative dir for denial URL
	char *preauthdir;					//@brief Notional relative dir for preauth URL
	int session_timeout;					//@brief Minutes of the default session length
	int preauth_idle_timeout;				//@brief Minutes a preauthenticated client will be kept in the system
	int auth_idle_timeout;					//@brief Minutes an authenticated client will be kept in the system
	int remotes_refresh_interval;				//@brief Minutes before remote files will be refreshed
	unsigned long long int remotes_last_refresh;		//@brief Time of last refresh of remote files
	int checkinterval;					//@brief Period the the client timeout check thread will run, in seconds
	int set_mss;						//@brief boolean, whether to set mss
	int mss_value;						//@brief int, mss value; <= 0 clamp to pmtu
	int rate_check_window;					//@brief window size in multiples of checkinterval for rate check moving average
	unsigned long long int download_rate;			//@brief Download rate, kb/s
	unsigned long long int upload_rate;			//@brief Upload rate, kb/s
	unsigned long long int download_bucket_ratio;		//@brief Allows control of download rate limit threshold overrun per client
	unsigned long long int upload_bucket_ratio;		//@brief Allows control of upload rate limit threshold overrun per client
	unsigned long long int max_upload_bucket_size;		//@brief control upload rate limiting packet loss at the expense of increased latency
	unsigned long long int max_download_bucket_size;	//@brief control download rate limiting packet loss at the expense of increased latency
	unsigned long long int download_quota;			//@brief Download quota, kB
	unsigned long long int upload_quota;			//@brief Upload quota, kB
	unsigned long long int fup_download_throttle_rate;	//@brief Fair Useage Policy Download throttle rate, kb/s, activated when quota exceeded
	unsigned long long int fup_upload_throttle_rate;	//@brief Fair Usage Policy Upload throttle rate, kb/s, activated when quota exceeded
	int download_unrestricted_bursting;			//@brief Enable/disable unrestriced bursting
	int upload_unrestricted_bursting;			//@brief Enable/disable unrestriced bursting
	int syslog_facility;					//@brief facility to use when using syslog for logging
	int macmechanism; 					//@brief mechanism wrt MAC addrs
	t_MAC *trustedmaclist;					//@brief list of trusted macs
	t_FASPARAM *fas_custom_parameters_list;			//@brief list of Custom FAS parameters
	t_FASVAR *fas_custom_variables_list;			//@brief list of Custom FAS variables
	t_FASIMG *fas_custom_images_list;			//@brief list of Custom FAS images
	t_FASFILE *fas_custom_files_list;			//@brief list of Custom FAS files
	char *custom_params;					//@brief FAS custom parameter string
	char *custom_vars;					//@brief FAS custom variable string
	char *custom_images;					//@brief FAS custom image string
	char *custom_files;					//@brief FAS custom file string
	unsigned int fw_mark_authenticated;			//@brief nftables mark for authenticated packets
	unsigned int fw_mark_auth_blocked;			//@brief nftables mark for auth_blocked packets
	char *authentication_mark;				//@brief Padded authentication mark
	unsigned int fw_mark_trusted;				//@brief nftables mark for trusted packets
	int ip6;						//@brief enable IPv6
	char *binauth;						//@brief external postauthentication program
	char *preauth;						//@brief external preauthentication program
	int lockfd;						//@brief ndsctl lockfile file descriptor
} s_config;

// @brief Get the current gateway configuration
s_config *config_get_config(void);

// @brief Initialise the conf system
void config_init(int argc, char **argv);
void parse_trusted_mac_list(const char[]);
void parse_fas_custom_parameters_list(const char[]);
void parse_fas_custom_variables_list(const char[]);
void parse_fas_custom_images_list(const char[]);
void parse_fas_custom_files_list(const char[]);

int is_trusted_mac(const char *mac);

int remove_from_trusted_mac_list(const char possiblemac[]);
int add_to_trusted_mac_list(const char possiblemac[]);

int check_ip_format(const char[]);
int check_mac_format(const char[]);

// config API, used in commandline.c
int set_debuglevel(const char[]);

#define LOCK_CONFIG() do { \
	debug(LOG_DEBUG, "Locking config"); \
	pthread_mutex_lock(&config_mutex); \
	debug(LOG_DEBUG, "Config locked"); \
} while (0)

#define UNLOCK_CONFIG() do { \
	debug(LOG_DEBUG, "Unlocking config"); \
	pthread_mutex_unlock(&config_mutex); \
	debug(LOG_DEBUG, "Config unlocked"); \
} while (0)

#endif // _CONF_H_

char *set_list_str(char *list, const char *default_list, char *debug_level);
